### ISSUE TITLE
Generic Types

### ISSUE THREAD

**Issue by Dominilk** [Lej77/cast_trait_object#1]
Hello! I have been trying to get this library working with traits that have generic types.
But for some reason, I can not really get it to work.

It would be awesome if you could provide me with an example!

Thanks,
Dominik

**Comment by Lej77** [comment]
Hi! Unfortunately this crate doesn't currently work with generic types. I should probably make that clearer in the documentation. (This is actually the one [todo item](https://github.com/Lej77/cast_trait_object/blob/541f512a1bdae4ba8aed71e70b6e4aee2caaae88/cast_trait_object_macros/src/lib.rs#L292) I didn't finish before publishing.)

A workaround that might work if you only use the generic type with some concrete types is to just manually implement the conversion for the types you care about. 
```rust
impl_dyn_cast!(Vec<u32> as MySuperTrait => MySubTrait); 
impl_dyn_cast!(Vec<bool> as MySuperTrait => MySubTrait);
//...
```

Another solution would be to just manually implement [`cast_trait_object::DynCast`](https://docs.rs/cast_trait_object/0.1.1/cast_trait_object/trait.DynCast.html) for any such generic types, this could work well if most types aren't generic but it is a bit of a hack.

A better solution would be if this crate had an alternative macro to the `impl_dyn_cast` macro that could be used on generic types (or it could just be another macro "arm" really but it would need to generate different code). The procedural attribute macros could probably automatically handle this difference. The [current macro](https://github.com/Lej77/cast_trait_object/blob/541f512a1bdae4ba8aed71e70b6e4aee2caaae88/cast_trait_object/src/lib.rs#L966-L991) tries to determine if the provided type implements the specified trait and generates different code based on the result. This is harder with generic types since they probably don't implement the trait unless their generic parameters are constrained in some way. Maybe we could just provide a where clause to the macro and check if the type implements the target trait when that where clause is used. That would probably work well enough. I'll try and see if I can make that work.

**Comment by Dominilk** [comment]
Thanks! Hope you are able to make it work :)

**Comment by Lej77** [comment]
I just released version `0.1.2` of this crate and it now has support for generic traits and types. I added a [`Generics`](https://docs.rs/cast_trait_object/0.1.2/cast_trait_object/#generics) section to the docs so you can check that out for some information about how it works and the limitations it currently has.

**Comment by Dominilk** [comment]
Awesome! I’ll close the issue now, thanks :)
